File: flink-scala-core/src/main/scala/com/atguigu/flink/day6/GenPeriodicWatermark.scala
package com.atguigu.flink.day6

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.{AssignerWithPeriodicWatermarks, KeyedProcessFunction}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.watermark.Watermark
import org.apache.flink.util.Collector

object GenPeriodicWatermark {

    def main(args: Array[String]): Unit = {
        val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
        env.setParallelism(1)
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
        env.getConfig.setAutoWatermarkInterval(5 * 1000L)

        env.socketTextStream("hadoop102", 9999, '\n')
                .map(line => {
                    val data: Array[String] = line.split(" ")
                    (data(0), data(1).toLong * 1000L)
                }).assignTimestampsAndWatermarks(
            new AssignerWithPeriodicWatermarks[(String, Long)] {

                val bound = 10 * 1000L
                var tsMax = Long.MinValue + bound + 1

                override def getCurrentWatermark: Watermark = {
                    val mark = tsMax - bound - 1
                    println("current :" + mark)
                    new Watermark(mark)
                }

                override def extractTimestamp(element: (String, Long), recordTimestamp: Long) = {
                    println("extract timestamp!")
                    tsMax = tsMax.max(element._2)
                    element._2
                }
            }
        ).keyBy(_._1)
                .process(new KeyedProcessFunction[String, (String, Long), String] {
                    override def processElement(value: (String, Long), ctx: KeyedProcessFunction[String, (String, Long), String]#Context, out: Collector[String]) = {
                        out.collect("current watermark :" + ctx.timerService().currentWatermark())
                    }
                }).print()

        env.execute()
    }

}

File: flink-java-core/src/main/java/com/atguigu/flink/day6/GenPeriodicWatermark.java
package com.atguigu.flink.day6;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.streaming.api.watermark.Watermark;
import org.apache.flink.util.Collector;

import javax.annotation.Nullable;

/**
 * @author ：hmz
 * @date ：Created in 2020/9/16 10:09
 */
public class GenPeriodicWatermark {

    public static void main(String[] args) throws Exception{
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.getConfig().setAutoWatermarkInterval(5 * 1000L);

        env.socketTextStream("hadoop102",9999)
                .map(new MapFunction<String, Tuple2<String,Long>>() {
                    @Override
                    public Tuple2<String, Long> map(String value) throws Exception {
                        String[] s = value.split(" ");
                        return Tuple2.of(s[0],Long.parseLong(s[1]) * 1000L);
                    }
                })
                .assignTimestampsAndWatermarks(
                        new AssignerWithPeriodicWatermarks<Tuple2<String, Long>>() {
                            private Long bound  = 10 * 1000L;
                            private Long tsMax = Long.MIN_VALUE + bound + 1;

                            @Override
                            public Watermark getCurrentWatermark() {
                                long mark = tsMax - bound - 1;
                                System.out.println("current auto:" + mark);
                                return new Watermark(mark);
                            }

                            @Override
                            public long extractTimestamp(Tuple2<String, Long> element, long recordTimestamp) {
                                System.out.println("extract timestamp!");
                                tsMax = Math.max(tsMax,element.f1);
                                return element.f1;
                            }
                        }
                )
                .keyBy(r->r.f0)
                .process(new KeyedProcessFunction<String, Tuple2<String, Long>, String>() {
                    @Override
                    public void processElement(Tuple2<String, Long> value, Context ctx, Collector<String> out) throws Exception {
                        out.collect("current " + ctx.timerService().currentWatermark());
                    }
                }).print();

        env.execute();
    }
}
